### PR TITLE
[feat] Add training-time validation support

### PR DESCRIPTION
**Pull request by WenzheWang**
## Summary

Implements the `EvaluateCallback._evaluate()` TODO stub to enable periodic validation during training. This addresses the existing placeholder in `veomni/trainer/callbacks/evaluate_callback.py` (line 37-39: `# TODO: implement evaluate`).

### New Files

| File | Description |
|------|-------------|
| `veomni/data/evaluator.py` | Core evaluator module with Registry pattern and distributed-aware aggregation. Supports 4 metrics: `loss`, `perplexity`, `accuracy`, `token_accuracy`. |
| `veomni/data/validation_loader.py` | Builds validation dataloader reusing existing `build_dataloader` infrastructure with deterministic eval settings (`shuffle=False`, `drop_last=False`, `dyn_bsz=False`). |
| `tests/data/test_evaluator.py` | pytest unit tests covering all evaluators, packed-batch boundaries, and distributed aggregation. |
| `docs/usage/training_with_validation.md` | User documentation with configuration guide and custom metric registration example. |
| `configs/examples/validation_example.yaml` | Configuration example. |

### Modified Files

| File | Changes |
|------|---------|
| `veomni/trainer/callbacks/evaluate_callback.py` | Implements `_ensure_built()` for lazy init, `_evaluate()` with `model.eval()`, `torch.no_grad()`, batch iteration, per-evaluator `aggregate()`, wandb logging, `model.train()` restore. Uses `inspect.signature()` to check `use_cache` support for DiT compatibility. |
| `veomni/data/__init__.py` | Appends evaluator exports (preserves all existing exports). |
| `veomni/arguments/arguments_types.py` | Adds `validation_metrics: List[str]` field to `TrainingArguments` (default: `["loss"]`). |

## Design Decisions

1. **Registry pattern**: Follows existing `DATA_TRANSFORM_REGISTRY` convention for extensibility
2. **Distributed-aware**: Per-rank partial sums aggregated via `dist.all_reduce()` 
3. **Lazy initialization**: Dataloader/evaluators built on first eval trigger, not at training start
4. **Backward compatible**: Default `validation_metrics=["loss"]`, eval disabled when `eval_steps=0` and `eval_epochs=0`
5. **Causal LM shift**: Evaluators apply standard `logits[:-1]` / `labels[1:]` shift
6. **ignore_index support**: `-100` labels automatically masked
7. **DiT compatibility**: Uses `inspect.signature()` to detect `use_cache` support, handles DDP-wrapped models via `model.module`

## Usage

```yaml
# In config YAML
train:
  eval_steps: 500        # Evaluate every 500 steps
  validation_metrics: ["loss", "perplexity", "accuracy"]

data:
  eval_path: /path/to/validation/data
```

## Testing

Verified on GPU (80GB) with PyTorch 2.3.1+cu118, Python 3.10.18:

- `verify_evaluators.py`: **7/7 PASSED** (including GPU large-scale test with batch=4, seq=512, vocab=32000)
- `pytest tests/data/test_evaluator.py`: **22/22 PASSED**

```
===== test session starts =====
platform linux -- Python 3.10.18, pytest-9.1.1, pluggy-1.6.0
collected 22 items

tests/data/test_evaluator.py ......................                     [100%]

===== 22 passed in 3.29s =====
```

Test coverage includes:
- Deterministic logits with exact metric value assertions
- Packed-batch boundary masking (IGNORE_INDEX at sample boundaries)
- Multi-rank distributed aggregation via real `dist.all_reduce` (gloo backend)
- Registry lookup and error handling

## Checklist

- [x] Apache 2.0 license headers on all new files
- [x] Follows existing code style (Registry pattern, `logging.get_logger`, `info_rank0`)
- [x] No new dependencies required (uses only `torch`, `torch.distributed`, `inspect`)
- [x] All existing exports preserved in `__init__.py`
- [x] Unit tests included and passing (deterministic, exact assertions)
- [x] Documentation included
- [x] Configuration example included

## Summary by CodeRabbit

* **New Features**
  * Added training-time validation with configurable evaluation intervals.
  * Added loss, perplexity, accuracy, and token accuracy metrics.
  * Added distributed metric aggregation and support for custom evaluators.
  * Added deterministic validation data loading and optional experiment tracking.
* **Documentation**
  * Added validation configuration examples and usage guidance.
* **Tests**
  * Added comprehensive coverage for evaluators, metric aggregation, batching, and distributed validation.

### CODE REVIEW

**Comment by WenZheWang** [comment]
@coderabbitai review

**Comment by WenZheWang** [comment]
@coderabbitai review

**Comment by WenzheWang** [comment]
@coderabbitai review
